src/plugins/hidsim/hid_injection.cpp

```diff
@@ -0,0 +1,765 @@
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include<fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <linux/input.h>
+#include <sys/time.h>
+#include <math.h>
+
+#include <libdrakvuf/libdrakvuf.h> /* eprint_current_time */
+#include "../private.h" /* PRINT_DEBUG */
+
+/* Hidsim specific includes */
+#include "keymap_evdev_to_qapi.h" /* Mapping evdev->qapi name */
+#include "qmp_connection.h"
+#include "hid_injection.h"
+
+/* Drakvuf HID template */
+#define DRAK_MAGIC_NUMBER 0xc4d2c1cb
+#define DRAK_MAGIC_STR 0x4b415244
+#define HEADER_LEN 0xC
+
+#define PPM_MAGIC_NUMBER "P6"
+
+/* Length of the QMP command buffer */
+#define CMD_BUF_LEN 0x400
+
+/* QAPI's button values */
+#define L_BTN_STR "left"
+#define M_BTN_STR "middle"
+#define R_BTN_STR "right"
+
+/* Throttle injection down to 50 microsec intervals */
+#define TIME_BIN 50
+
+struct dimensions
+{
+    int width;      /* display width */
+    int height;     /* display height */
+    float dx;       /* one pixel equivalent on x-axis */
+    float dy;       /* one pixel equivalent on y-axis */
+};
+
+/* Keep track of the current cursor position */
+int new_x = 1 << 14;
+int new_y = 1 << 14;
+
+static int dump_screen(qmp_connection* qc, const char* path)
+{
+    char cmd[0x200];
+    snprintf(cmd, 0x200,
+        QMP_SCREEN_DUMP_FMT_STR,
+        path);
+    char* out = NULL;
+
+    qmp_communicate(qc, cmd, &out);
+
+    if (strcmp(out, QMP_SUCCESS) != 0)
+    {
+        fprintf(stderr, "[HIDSIM]: Error dumping screen - %s\n", out);
+    }
+    return 0;
+}
+
+/* For absolute pointing devices 2^15 is used to specify maximum */
+static float calculate_pixel_unit(int u)
+{
+    return (float)(1<<15)/u;
+}
+
+/* Retrieves display dimensions for coordination of mouse movements */
+static int get_display_dimensions(qmp_connection* qc, struct dimensions* dims)
```

Comment:
I've found an alternative way to implement this:
We can call `qom-list /machine/unattached` to query devices emulated by QEMU.
In my case there's `device[6] (child<VGA>)`. Now we can call:
`qom-get /machine/unattached/device[6]/xres` to get width, and similarly `yres` to get height of the display.

What do you think?

Reply:
I evaluated the retrieval via QEMU's device model as well, but in my case it is -- somehow -- not that easy. Looking in "my" device tree yields the following content regarding vga:  
```
/machine/peripheral-anon/device[0]
  type: VGA (string)
  parent_bus: /machine/i440fx/pci.0 (link<bus>)
  realized: True (bool)
  <snip>
  ymax: 0 (uint32)
  yres: 0 (uint32)
  xmax: 0 (uint32)
  edid: False (bool)
  xres: 0 (uint32)
  <snip>
```
Therefore the utilization of screendump seemed to work best. 
Did I oversee something or are my thought erroneous?

Reply:
Ah yes, you're right. I was lazy and used "normal" QEMU instead. Seems like Xen doesn't enable edid (?).
Oh well.

Reply:
Yes, looks like it is disable by default, see https://xenbits.xen.org/docs/unstable/misc/xen-command-line.html#edid-x86
But don't know, if this is the reason, for not being able to retrieve this info.

Reply:
The xen command line option has nothing to do what option of qemu gets enabled. Use the xl config option `device_model_args = []` to pass addition arguments to QEMU if you need to.

Reply:
> The xen command line option has nothing to do what option of qemu gets enabled. Use the xl config option `device_model_args = []` to pass addition arguments to QEMU if you need to.

Thanks for clarification, @tklengyel! 

If everybody would be okay with the current solution using the qmp-`screendump`-command to retrieve the display dimensions, I would leave it as it is, since it is a generic solution, which is decoupled of the specification of a specific `device_model` IMHO. What do you think? 

Furthermore, is there anything, that I should change until the PR is ready for merging?  
The "default"-build still fails, but by browsing the UI on Jenkins I can't derive the reason for that, do you have a hint, where to look for the problem?